### ISSUE TITLE
[FastPitch 1.1/PyTorch] Advice/best practices for good alignment when fine-tuning

### ISSUE THREAD

**Issue by DanRuta** [NVIDIA/DeepLearningExamples#1004]
Related to **FastPitch 1.1/PyTorch** 

Hi @alancucki @rafaelvalle. I've been experimenting with the FastPitch 1.1 update incorporating rad-tts into fastpitch, since the commit a while back. The alignment mechanism is amazing, and enables some great things (eg s2s).

However, I'm unfortunately having some issues with the convergence on some datasets with this approach, compared to 1.0. I've successfully converged an LJSpeech model quite well, but when fine-tuning a pre-trained model (like the one provided), it seems that the alignment is having some real trouble converging.

I have tried it on 4 datasets so far - 2 male, 2 female. I did get something ALMOST good with one of the female datasets (~9h), but they all converge to a fairly high KL loss (>=0.85 at 40k-120k its, compared to less than 0.35 where I stopped LJ at 25k its). I added soft and hard alignment plots to the logs, and they resemble plots c) and d) from Fig 2 in the RAD-TTS paper. I noticed also in Figure 2 from "One TTS Alignment To Rule Them All", that the convergence speed was lower using RAD-TTS with FastPitch (compared to Tacotron2 durs), before they arrived at a similar point - could this be exacerbated by smaller datasets? 

I have tried experimenting with resuming the LJ optimizer (from the model I trained myself) as well as one newly initialized (from the provided LJ model), with and without including the KL weight warm-up stage. I also tried with and without arpabet, with and without energy conditioning, and also several tweaks to lr scheduling, and other such things, but I can never get anything as good as LJ (in the KL loss at least).

When running inference, the sentence composition quality varies between datasets, ranging from missing letters to missing words, and for the smaller datasets, quite difficult to understand speech, spoken very fast.

The same datasets worked very well in the previous Tacotron2+FastPitch set-up, so I'm confident that the data quality is high. Have you by any chance had any successes yourselves with something other than LJ? And would you have any tips/advice for how to better converge the alignment on smaller datasets (with transfer learning)? 

Thank you for all your great work!

**Comment by rafaelvalle** [comment]
fastpitch2 already has the alignment method used in radtts and i think there's a verion on nvidia's nemo repo. 
adrian can you point you the code.

**Comment by DanRuta** [comment]
Is this different from FastPitch 1.1? I assumed that was the final version where RAD-TTS was added - exciting, if I was wrong. 

I see a link on nemo (https://ngc.nvidia.com/catalog/models/nvidia:tao:speechsynthesis_english_fastpitch) which mentions FastPitch 2 (btw, the arxiv link points to 1.0), but is this different from the code in this repo?

**Comment by zsh01** [comment]
Hi @DanRuta! Sorry for a laggy reply.

> The same datasets worked very well in the previous Tacotron2+FastPitch set-up

Did those datasets have their own Tacotron 2 models, or did you predict their durations with LJ Tacotron 2?

> I have tried it on 4 datasets so far - 2 male, 2 female. I did get something ALMOST good with one of the female datasets

Pretty much all datasets I've tried worked without issues in training from scratch. When the dataset is smaller, I had good results combining few speakers and using speaker embeddings (there is support for that in the code already). Note that kl warmup schedule should then be re-scaled proportionally.

I assume you're fine-tuning instead of training from scratch due to resource constraints. It is a long shot, but I'd try to add the smaller dataset (say 9 h) to LJSpeech-1.1 and fine-tune on both of them. Speaker embedding for LJ could be initially a vector of zeros (it's additive anyway). If you have a tiny dataset, you can quickly increase its size relative to LJ just concatenating a couple of times its filelist. Also, remember that pitch normalizing constants (mean and std) are by default set to those of LJ.

> Is this different from FastPitch 1.1? 

No, I think @rafaelvalle meant v1.1. We've reserved the v2 version number for a more substantial update, but it seems that it already caught on.

btw Thanks for the detailed feedback, it's really valuable! Let me know how if adding speakers for finetuning fixes thigs.

**Comment by alancucki** [comment]
Hi @DanRuta! Sorry for a laggy reply.

> The same datasets worked very well in the previous Tacotron2+FastPitch set-up

Did those datasets have their own Tacotron 2 models, or did you predict their durations with LJ Tacotron 2?

> I have tried it on 4 datasets so far - 2 male, 2 female. I did get something ALMOST good with one of the female datasets

Pretty much all datasets I've tried worked without issues in training from scratch. When the dataset is smaller, I had good results combining few speakers and using speaker embeddings (there is support for that in the code already). Note that kl warmup schedule should then be re-scaled proportionally.

I assume you're fine-tuning instead of training from scratch due to resource constraints. It is a long shot, but I'd try to add the smaller dataset (say 9 h) to LJSpeech-1.1 and fine-tune on both of them. Speaker embedding for LJ could be initially a vector of zeros (it's additive anyway). If you have a tiny dataset, you can quickly increase its size relative to LJ just concatenating a couple of times its filelist. Also, remember that pitch normalizing constants (mean and std) are by default set to those of LJ.

> Is this different from FastPitch 1.1?

No, I think @rafaelvalle meant v1.1. We've reserved the v2 version number for a more substantial update, but it seems that it already caught on.

btw Thanks for the detailed feedback, it's really valuable! Let me know how if adding speakers for finetuning fixes thigs.

**Comment by alancucki** [comment]
@DanRuta , that is quite a collection of Tacotron and FastPItch models :)

> I've tweaked a couple of things to allow non-binary arpabet probability, and I'm now training models with 0.5 prob of text/phoneme input.

Does it work equally well with FastP 1.0 and 1.1?

> The LJ default values in 1.1 are somewhat similar to the values in 1.0, so I'm wondering if you also used the same process.

tbh I haven't seen much difference when tweaking mean/std values. I guess getting right normalization would be more important with a multispeaker model.

> Finally, I know with 1.0 that there was no recommended way to do early stopping, but rather to intermittently check the quality in inference once the training loss flattens. Is this still the case?

Unfortunately yes. And it is a bit tricky - I have a strong feeling that, when you're trying to do early stopping, the optimal epoch depends on which vocoder will be applied later on.

**Comment by dsplog** [comment]
@DanRuta  : nice and helpful to share your observations. can you please comment on how the model was extended for speech-to-speech synthesis.

**Comment by longjoke** [comment]
@DanRuta did you train a LJ model with 0.5 probability and then use transfer learning for the smaller datasets or are you using the NVIDIA model?

**Comment by DanRuta** [comment]
@dsplog No extension needed. Given a transcript for the input audio (gt, or ASR inferred), the RAD-TTS alignment in a FastPitch model trained on a reference voice A can infer durations for every symbol. Those durations can then be used to extract pitch and energy values from the reference audio. Then, using the durations, pitch, and energy values, you can perform partial inference with a FastPitch model, for your target voice B. You can use/see an implementation of this [here](https://github.com/DanRuta/xVA-Synth), or see a video of it [here](https://www.youtube.com/watch?v=W-9SFoNuTtM).

@longjoke I fine-tuned the NVIDIA LJ model using 0.5 ARPAbet probability first, before transfer learning on my own datasets.

**Comment by adrian585** [comment]
Hi @alancucki @rafaelvalle, @DanRuta . 

I want to retrain "FastPitch1.1" to another voice. Wrote an algorithm that calculates mean and std. To make sure the algorithm is correct, I decided to calculate these values for the LJ Speech Dataset. 
I got the following:
mean = 212.3237 (your value is 214.72203)
std = 68.5746 (your value is 65.72038)

steps I took:
1. ran module "prepare_dataset.py" with all 13100 LJ dataset's wavs. Thus, I got all pitch files.
2. executed torch.load() for each pitch-file(13100 files) and removed all zeros from tensors.
3. Combined all tensors into one.
4. Executed torch.mean() and torch.std() for this tensor.

Why are my mean and std so close to yours, but still different?
I would be very grateful if you point out my mistake. Thank you in advance.
